general/development/policies/routing/index.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Routing Policies
+tags:
+  - Processes
+  - Core development
+  - Policy
+  - Routing
+  - Pages
+---
+
+This document describes the policies for route naming in Moodle.
+
+Routes are a powerful feature introduced in Moodle 4.5, and further extended in Moodle 5.0 which allow parts of Moodle, including plugins, to generate content without an explicit PHP endpoint file directly hit by users.
+
+See the [Routing system documentation](/docs/apis/subsystems/routing) for further information.
+
+Routes are currently used in the following places:
+
+- The REST API
+- End user content (pages)
+- Shims to redirect legacy locations to new pages (shims)
+
+Part of the URL used to access these pages is automatically generated by the component that the route is a part of, but the remaining part is generated by the developer.
+
+To ensure consistency across Moodle, and to mitigate the risk of route shadowing, Moodle follows some rules for route paths.
+
+Routes are typically comprised of:
+
+- the component prefix
+- an _entity_ type, for example `course`
+- an entity _identifier_, for example `948`
+- an _action_ on that entity, for example `view`
+
+Some of these parts may not be required, depending on the component and route type.
+
+## The Component {#component-naming}
+
+Where a component is present in the URL, the value is automatically determined from the location of the class defining the route.
+
+:::tip Component Appearance
+
+Where the component is a plugin, the full Frankenstyle Plugin Name is used, for example `mod_assign`.
+
+Where the component is a core subsystem such as `core_course` then the `core_` prefix is removed, for example `course`.
+
+:::
+
+## Entity types
+
+In many cases the component will be the entity, for example in the `core_course` component, the primary entity is the course. This is referred to as the _default_ entity.
+
+The following are examples of a default entity:
+
+``` title="The core_course subsystem has a default entity of 'course'"
+/rest/api/v2/course/<courseid>
+/course/<courseid>/view
+```
+
+The following are examples of a non-default entity:
+
+``` title="The core subsystem does not have any default entity"
+/rest/api/v2/core/templates/<theme>/<templateName>
+/rest/api/v2/core/strings/<language>/<stringComponent>/<stringIdentifier>
+```
+
+:::danger Default and non-default entities
+
+The use of both a _default_ entity _and_ a non-default entity is not officially supported.
+
+:::
+
+:::tip Singular or Plural
+
+Where present, entities should always be in the plural form.
```

Comment:
I agree about [plural form](https://tracker.moodle.org/browse/MDL-76834?focusedId=1122537&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1122537)! Should examples for course entity be changed accordingly?

Reply:
We don't need to strictly follow the subsystems names here. 
However, in the example of course, `/courses/<courseid>/view` sounds not correct, as you are only going to ever look at a singular course. So I think in this case singular is the way to go. 
Not sure how we make the docs clearer though.

Reply:
At the moment the component resolution is automatic based on the location of the class and we automatically normalise it.

If we want to do something different then we'll have to work out the best way of doing so and in which situations.

Reply:
> However, in the example of course, `/courses/<courseid>/view` sounds not correct, as you are only going to ever look at a singular course. So I think in this case singular is the way to go.

Yeah, it is sounds not correct in individual scenario, unless one see it as query "from courses select <courseid>". Thinking in broader context, what will be the endpoint for querying list of courses? Different `/courses` path or `/course/list` action? Something better to consider while we are at early stage I guess :)

With regard to this "tip", to avoid contradition for now we might say "Where possible, entities should always be in the plural form."

PS In Martin Fowler [REST maturity model](https://martinfowler.com/articles/richardsonMaturityModel.html) article all entities are plural.

Reply:
Andrew's point is a good one that resolution is based on automatic resolution. Messing around with that feels like it may add more complexity than it's worth.
Also, thanks for the extract context Ruslan